Added a new file - k3k.adoc (#11)
rancher/k3k-product-docs: versions/latest/modules/en/pages/k3k.adoc

= k3k: Kubernetes in Kubernetes

K3k, Kubernetes in Kubernetes, is a tool that empowers you to create and manage isolated K3s clusters within your existing Kubernetes environment. It enables efficient multi-tenancy, streamlined experimentation, and robust resource isolation, minimizing infrastructure costs by allowing you to run multiple lightweight Kubernetes clusters on the same physical host. K3k offers both "shared" mode, optimizing resource utilization, and "virtual" mode, providing complete isolation with dedicated K3s server pods. This allows you to access a full Kubernetes experience without the overhead of managing separate physical resources.

> K3k integrates seamlessly with Rancher for simplified management of your embedded clusters.

Reviewer: I don't think the integrates needs to be over-qualified.
```suggestion
K3k integrates with Rancher for simplified management of your embedded clusters.
```

== Features and Benefits

> * *Resource Isolation*: Ensure workload isolation and prevent resource contention between teams or applications. K3k allows you to define resource limits and quotas for each embedded cluster, guaranteeing that one team's workloads won't impact another's performance.

Reviewer: ```suggestion
* *Resource Isolation*: Enables workload isolation and prevents resource contention between teams or applications. K3k allows you to define resource limits and quotas for each embedded cluster, so that one team's workloads won't impact another's performance.
```

> * *Simplified Multi-Tenancy*: Easily create dedicated Kubernetes environments for different users or projects, simplifying access control and management. Provide each team with their own isolated cluster, complete with its own namespaces, RBAC, and resource quotas, without the complexity of managing multiple physical clusters.

Reviewer: ```suggestion
* *Simplified Multi-Tenancy*: Easily create dedicated Kubernetes environments for different users or projects, simplifying access control and management. Provide each team with their own cluster, complete with its own namespaces, RBAC, and resource quotas, without the complexity of managing multiple physical clusters.
```

> * *Lightweight and Fast*: Leverage the lightweight nature of K3s to spin up and tear down clusters quickly, accelerating development and testing cycles. Spin up a new K3k cluster in seconds, test your application in a clean environment, and tear it down just as quickly, streamlining your CI/CD pipeline.

Reviewer: ```suggestion
* *Lightweight and Fast*: Use the lightweight nature of K3s to spin up and tear down clusters quickly, accelerating development and testing cycles. Spin up a new K3k cluster in seconds, test your application in a clean environment, and tear it down just as quickly, streamlining your CI/CD pipeline.
```

* *Optimized Resource Utilization (Shared Mode)*: Maximize your infrastructure investment by running multiple K3s clusters on the same physical host. K3k's shared mode allows you to efficiently share underlying resources, reducing overhead and minimizing costs.

* *Complete Isolation (Virtual Mode)*: For enhanced security and isolation, K3k's virtual mode provides dedicated K3s server pods for each embedded cluster. This ensures complete separation of workloads and eliminates any potential resource contention or security risks.

> * *Rancher Integration*: Simplify the management of your K3k clusters with Rancher. Leverage Rancher's intuitive UI and powerful features to monitor, manage, and scale your embedded clusters with ease.

Reviewer: ```suggestion
* *Rancher Integration*: Simplify the management of your K3k clusters with Rancher. Use Rancher's UI and  features to monitor, manage, and scale your embedded clusters.
```

* An existing https://documentation.suse.com/cloudnative/rke2/latest/en/install/quickstart.html[RKE2] Kubernetes cluster (recommended).
* A configured storage provider with a default storage class.

NOTE: If you do not have a storage provider, you can configure the cluster to use ephemeral or static storage. Please consult the xref:advanced-usage.adoc#_using_the_cli[k3kcli Advanced Usage] for instructions on using these options.

=== Install the K3k controller

. Add the K3k Helm repository:
+
[,bash]
----
helm repo add k3k https://rancher.github.io/k3k
helm repo update
----

. Install the K3k controller:
+
[,bash]
----
helm install --namespace k3k-system --create-namespace k3k k3k/k3k
----

=== Install the k3kcli

The `k3kcli` provides a quick and easy way to create K3k clusters and automatically exposes them via a kubeconfig.

To install it, simply download the latest available version for your architecture from the GitHub Releases page.

For example, you can download the Linux amd64 version with:

[,bash]
----
wget -qO k3kcli https://github.com/rancher/k3k/releases/download/{latest-version}/k3kcli-linux-amd64 && \
  chmod +x k3kcli && \
  sudo mv k3kcli /usr/local/bin
----

> You should now be able to run:

Reviewer: ```suggestion
Now run:
```

INFO[0073] You can start using the cluster with: 

        export KUBECONFIG=/my/current/directory/mycluster-kubeconfig.yaml
        kubectl cluster-info  
----

After exporting the generated kubeconfig, you should be able to reach your Kubernetes cluster:

[,bash]
----
export KUBECONFIG=/my/current/directory/mycluster-kubeconfig.yaml
kubectl get nodes
kubectl get pods -A
----

You can also directly create a Cluster resource in some namespace, to create a K3k cluster:

[,bash]
----
kubectl apply -f - <<EOF
apiVersion: k3k.io/v1alpha1
kind: Cluster
metadata:
  name: mycluster
  namespace: k3k-mycluster
EOF
----

and use the `k3kcli` to retrieve the kubeconfig:

[,bash]
----
k3kcli kubeconfig generate --namespace k3k-mycluster --name mycluster 
----

=== Deleting a K3k Cluster

To delete a K3k cluster, use the following command:

[,bash]
----
k3kcli cluster delete mycluster
----

== Architecture

For a detailed explanation of the `K3k` architecture, please refer to the xref:architecture.adoc[Architecture documentation].

== Advanced Usage

> For more in-depth examples and information on advanced K3k usage, including details on shared vs. virtual modes, resource management, and other configuration options, please see the xref:advanced-usage.adoc[Advanced Usage documentation].

Reviewer: ```suggestion
For more in-depth examples and information on advanced K3k usage, including details on the differences between shared and virtual modes, resource management, and other configuration options, please see the xref:advanced-usage.adoc[Advanced Usage documentation].
```

> = k3k: Kubernetes in Kubernetes

Reviewer 2: Please add revdate

Reviewer: I don't think it can be resolved until the actual revdate has been added.

K3k, Kubernetes in Kubernetes, is a tool that empowers you to create and manage isolated K3s clusters within your existing Kubernetes environment. It enables efficient multi-tenancy, streamlined experimentation, and robust resource isolation, minimizing infrastructure costs by allowing you to run multiple lightweight Kubernetes clusters on the same physical host. K3k offers both "shared" mode, optimizing resource utilization, and "virtual" mode, providing complete isolation with dedicated K3s server pods. This allows you to access a full Kubernetes experience without the overhead of managing separate physical resources.

K3k integrates seamlessly with Rancher for simplified management of your embedded clusters.

*Experimental Tool*

> This project is still under development and is considered experimental. It may have limitations, bugs, or changes. Please use with caution and report any issues you encounter. We appreciate your feedback as we continue to refine and improve this tool.

Reviewer 3: Assuming we're publishing to documentation.suse.com for the first time when SUSE® Virtual Clusters is launched, this doesn't like it would be accurate at that point. Can you double check.

Author: Removed.

* *Optimized Resource Utilization (Shared Mode)*: Maximize your infrastructure investment by running multiple K3s clusters on the same physical host. K3k's shared mode allows you to efficiently share underlying resources, reducing overhead and minimizing costs.

* *Complete Isolation (Virtual Mode)*: For enhanced security and isolation, K3k's virtual mode provides dedicated K3s server pods for each embedded cluster. This ensures complete separation of workloads and eliminates any potential resource contention or security risks.

* *Rancher Integration*: Simplify the management of your K3k clusters with Rancher. Leverage Rancher's intuitive UI and powerful features to monitor, manage, and scale your embedded clusters with ease.

== Installation

This section provides instructions on how to install K3k and the k3kcli.

=== Prerequisites

* https://helm.sh/[Helm] must be installed to use the charts. Please refer to https://helm.sh/docs/[Helm's documentation] to get started.

> * An existing https://docs.rke2.io/install/quickstart[RKE2] Kubernetes cluster (recommended).

Reviewer 3: ```suggestion
* An existing https://documentation.suse.com/cloudnative/rke2/latest/en/install/quickstart.html[RKE2] Kubernetes cluster (recommended).
```

We should link to other Prime/product docs when it's available.

* A configured storage provider with a default storage class.

> Note: If you do not have a storage provider, you can configure the cluster to use ephemeral or static storage. Please consult the https://github.com/rancher/k3k/blob/main/docs/advanced-usage.md#using-the-cli[k3kcli advance usage] for instructions on using these options.

Reviewer 2: ```suggestion
NOTE: If you do not have a storage provider, you can configure the cluster to use ephemeral or static storage. Please consult the https://github.com/rancher/k3k/blob/main/docs/advanced-usage.md#using-the-cli[k3kcli advance usage] for instructions on using these options.
```

> Note: If you do not have a storage provider, you can configure the cluster to use ephemeral or static storage. Please consult the https://github.com/rancher/k3k/blob/main/docs/advanced-usage.md#using-the-cli[k3kcli advance usage] for instructions on using these options.

Reviewer 3: ```suggestion
Note: If you do not have a storage provider, you can configure the cluster to use ephemeral or static storage. Please consult the xref:advanced-usage.adoc#_using_the_cli[k3kcli Advanced Usage] for instructions on using these options.
```

That file has been migrated to this repo so we shouldn't link to that k3k GitHub repo.

=== Install the K3k controller

. Add the K3k Helm repository:
+
[,bash]
----
helm repo add k3k https://rancher.github.io/k3k
helm repo update
----

. Install the K3k controller:
+
[,bash]
----
helm install --namespace k3k-system --create-namespace k3k k3k/k3k
----

> NOTE: K3k is currently under development. We recommend using the latest released version when possible.

Reviewer 3: Assuming we're publishing to documentation.suse.com for the first time when SUSE® Virtual Clusters is launched, this doesn't like it would be accurate at that point. Can you double check.

Author: Removed.

=== Install the k3kcli

The `k3kcli` provides a quick and easy way to create K3k clusters and automatically exposes them via a kubeconfig.

To install it, simply download the latest available version for your architecture from the GitHub Releases page.

For example, you can download the Linux amd64 version with:

[,bash]
----

> wget -qO k3kcli https://github.com/rancher/k3k/releases/download/v0.3.4/k3kcli-linux-amd64 && \

Reviewer 3: Maybe use an attribute instead of hardcoding v0.3.4 here and only line 76. The version will be something different at launch and we don't want to list a version that's not GA.

Author: The PM confirmed that the version will be 1.0.0 at the time of launch. I'll use the attribute.

INFO[0000] Extracting Kubeconfig for [mycluster] cluster 
INFO[0000] waiting for cluster to be available..        
INFO[0073] certificate CN=system:admin,O=system:masters signed by CN=k3s-client-ca@1738746570: notBefore=2025-02-05 09:09:30 +0000 UTC notAfter=2026-02-05 09:10:42 +0000 UTC 
INFO[0073] You can start using the cluster with: 

        export KUBECONFIG=/my/current/directory/mycluster-kubeconfig.yaml
        kubectl cluster-info  
----

After exporting the generated kubeconfig, you should be able to reach your Kubernetes cluster:

[,bash]
----
export KUBECONFIG=/my/current/directory/mycluster-kubeconfig.yaml
kubectl get nodes
kubectl get pods -A
----

You can also directly create a Cluster resource in some namespace, to create a K3k cluster:

[,bash]
----
kubectl apply -f - <<EOF
apiVersion: k3k.io/v1alpha1
kind: Cluster
metadata:
  name: mycluster
  namespace: k3k-mycluster
EOF
----

and use the `k3kcli` to retrieve the kubeconfig:

[,bash]
----
k3kcli kubeconfig generate --namespace k3k-mycluster --name mycluster 
----

=== Deleting a K3k Cluster

To delete a K3k cluster, use the following command:

[,bash]
----
k3kcli cluster delete mycluster
----

== Architecture

> For a detailed explanation of the `K3k` architecture, please refer to the xref:./architecture.adoc[Architecture documentation].

Reviewer 2: ```suggestion
For a detailed explanation of the `K3k` architecture, please refer to the xref:architecture.adoc[Architecture documentation].
```

== Advanced Usage

> For more in-depth examples and information on advanced K3k usage, including details on shared vs. virtual modes, resource management, and other configuration options, please see the xref:./advanced-usage.adoc[Advanced Usage documentation].

Reviewer 2: ```suggestion
For more in-depth examples and information on advanced K3k usage, including details on shared vs. virtual modes, resource management, and other configuration options, please see the xref:advanced-usage.adoc[Advanced Usage documentation].
```

== Development

> If you're interested in building K3k from source or contributing to the project, please refer to the xref:./development.adoc[Development documentation].

Reviewer 2: ```suggestion
If you're interested in building K3k from source or contributing to the project, please refer to the xref:development.adoc[Development documentation].
```

> == License

Reviewer 3: Is this license about the docs or k3k/SUSE Virtual Cluster itself?

Author: Apparently, most of such similar disclosures are present due to copy paste mistakes from another source. So, they asked to remove these statements entirely. 

That being said, the product manager wasn't sure and speculated that the license might be referring to the entire documentation.

> Copyright (c) 2014-2025 https://www.rancher.com/[SUSE]

Reviewer 3: Since the link text is SUSE shouldn't the URL be to SUSE as well?

The date range doesn't look correct either since neither k3k/SUSE Virtual Cluster nor these docs existed in 2014.

Author: Removed entirely. PM said it's apparently a copy paste mistake from their side.
